operator/pkg/client/client_factory.go

```diff
@@ -0,0 +1,370 @@
+package client
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/kyma-project/module-manager/operator/pkg/util"
+	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/kube"
+	"helm.sh/helm/v3/pkg/storage"
+	"helm.sh/helm/v3/pkg/storage/driver"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/cli-runtime/pkg/resource"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/discovery/cached/memory"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/kubectl/pkg/util/openapi"
+	openapivalidation "k8s.io/kubectl/pkg/util/openapi/validation"
+	"k8s.io/kubectl/pkg/validation"
+)
+
+const (
+	apis = "/apis"
+	api  = "/api"
+)
+
+// SingletonClients serves as a single-minded client interface that combines
+// all kubernetes Client APIs (Helm, Kustomize, Kubernetes, Client-Go) under the hood.
+// It offers a simple initialization lifecycle during creation, but delegates all
+// heavy duty work to deferred discovery logic and a single http client
+// as well as a client cache to support GV-based clients.
+type SingletonClients struct {
+	httpClient *http.Client
+
+	// the original config used for all clients
+	config *rest.Config
+
+	// discovery client, used for dynamic clients and GVK discovery
+	discoveryClient     discovery.CachedDiscoveryInterface
+	discoveryRESTMapper meta.ResettableRESTMapper
+	// expander for GVK and REST expansion from discovery client
+	discoveryShortcutExpander meta.RESTMapper
+
+	// kubernetes client
+	kubernetesClient *kubernetes.Clientset
+	dynamicClient    dynamic.Interface
+
+	// helm client with factory delegating to other clients
+	helmClient *kube.Client
+	install    *action.Install
+
+	// OpenAPI document parser singleton
+	openAPIParser *openapi.CachedOpenAPIParser
+
+	// OpenAPI document getter singleton
+	openAPIGetter *openapi.CachedOpenAPIGetter
+
+	// GVK based structured Client Cache
+	structuredSyncLock        sync.Mutex
+	structuredRestClientCache map[string]resource.RESTClient
+
+	// GVK based unstructured Client Cache
+	unstructuredSyncLock        sync.Mutex
+	unstructuredRestClientCache map[string]resource.RESTClient
+}
+
+var _ kube.Factory = &SingletonClients{}
```

Comment:
A comment here perhaps why it should be comply to the interface?

Reply:
This is regular Go imperative and I do not think it really needs an explanation. What would the benefit be of this comment?